THRIFT-3773 Swift 3 Native Library

apocolipse opened the pull request:
Here is my implementation of the Thrift library for Swift 3.0, pure swift.  
There is no Objective-C requirement here. 
For implementation and design details, please refer to comments here:
https://issues.apache.org/jira/browse/THRIFT-3773?jql=text%20~%20%22swift%22
And to the README here:
https://github.com/apocolipse/thrift/tree/master/lib/swift

fumoboy007 commented on lib/swift/Sources/TApplicationError.swift:
```diff
@@ -0,0 +1,157 @@
+  }
+  
+  public var hashValue: Int {
+    return error.thriftErrorCode + (message?.hashValue ?? 0)
```
`&+`?

fumoboy007 commented on lib/swift/Sources/TList.swift:
```diff
@@ -0,0 +1,135 @@
+    let prime = 31
+    var result = 1
+    for element in storage {
+      result = prime * result + element.hashValue
```
`&*`, `&+`?

fumoboy007 commented on lib/swift/Sources/TMap.swift:
```diff
@@ -0,0 +1,182 @@
+    let prime = 31
+    var result = 1
+    for (key, value) in storage {
+      result = prime * result + key.hashValue
```
`&*`, `&+`?

fumoboy007 commented on lib/swift/Sources/TSet.swift:
```diff
@@ -0,0 +1,175 @@
+    let prime = 31
+    var result = 1
+    for element in storage {
+      result = prime * result + element.hashValue
```
`&*`, `&+`?

fumoboy007 commented on lib/swift/Sources/TStruct.swift:
```diff
@@ -0,0 +1,100 @@
+    let prime = 31
+    var result = 1
+    self.forEach { _, value, _ in
+      result = prime * result + (value.hashValue)
```
`&*`, `&+`?

apocolipse commented:
@fumoboy007 thanks, thought i had the overflow ones in there, will add them back in :)

fumoboy007 commented on lib/swift/Sources/LinuxHelper.swift:
```diff
@@ -0,0 +1,42 @@
+extension URLSession {
+  // Current one uses NSURLRequest which doesn't currently bridge
+  @discardableResult
+  open func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
```
Not necessary anymore I think: https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/NSURLSession/NSURLSession.swift#L423

apocolipse replied:
@fumoboy007 Which tag was that added in?  I'm attempting to have this match up with GM (its been tricky the past few weeks -_-)
I don't mind floating it since on the Darwin side, this stuff is all good to go where its hard to update swift, whereas the linux side its a little easier to update Swift to snapshots to ensure a solid library build.

fumoboy007 replied:
I’m not too sure how the tagging/branching works in the Swift repo, but I see that the [commit](https://github.com/apple/swift-corelibs-foundation/commit/cffa65a1f933ddb29b7d3d21cd1756caeddc238f) is in `swift-3.0-branch` (as well as `master`) and is included in the `swift-3.0-GM-CANDIDATE` tag.

ChristopherRogers commented:
I noticed you haven't special cased unions (they'll just be generated as structs.) Have you looked into adding support for this? I'd love to be able to use Swift's enums with Thrift's union types.

apocolipse commented:
@ChristopherRogers Something like this?
```
union AUnion {
  1: string result,
  2: CustomException error
}
```
```swift
enum AUnion {
  case result(val: String)
  case error(val: CustomException)
}
```

apocolipse commented:
That'll require generation of both read and write serialization code (which I try to abstract much of to the library), but its doable I suppose.

nsuke commented:
There's a substantial amount of code duplication with existing swift generator.
Do you have any idea as to reducing them ?
As I've never seen anyone advocating the current practice of "single big file per language", we could, for example, extract common part to another file or even introduce a swift sub directory etc.

Internal structures aside, doesn't it make more sense to have a sub-option like --gen swift:3 rather than a separate language option ?

CC: @kdubb for review and/or comments.

apocolipse commented:
@nsuke I considered that when starting out, however the differences between Swift 2 and 3 are pretty dramatic that makes shared code minimal (i.e. everything in this library is Value typed, compared to the Swift/Cocoa generator).
Conceptually, this /is/ the `Swift` generator, and the other one is technically Cocoa Swift.
Additionally, due to limitations with SPM, sharing code might not be the most sane option.
One key difference with this and the existing Swift library, this library has no intention of Objective-C interop, it is a pure Swift approach.  

We could argue for moving the old Swift generator to be used like `--gen cocoa:swift` which would be more appropriate in my opinion.

As far as sharing code between them goes, the amount that is the same should be negligable, but packaging is the biggest concern.  Cocoapods isn't overly friendly with picking/choosing source files from different locations in the file heirarchy, and SPM isn't capable of pointing at sources not in `Sources` at all (hence the need for a separate repo for the time being to leverage it with SPM).

apocolipse commented:
@ChristopherRogers @fumoboy007 I implemented union support with enums, here's an example of generated code:
```
union TestUnion {
  1: i32 data,
  2: string message,
  3: bool success
}
```

```swift
public enum TestUnion {
  case data(val: Int32)
  case message(val: String)
  case success(val: Bool)
}
public func ==(lhs: TestUnion, rhs: TestUnion) -> Bool {
  return {
    switch (lhs, rhs) {
    case (.data(let lval), .data(let rval)): return lval == rval
    case (.message(let lval), .message(let rval)): return lval == rval
    case (.success(let lval), .success(let rval)): return lval == rval
    default: return false
    }
  }()
}

extension TestUnion : CustomStringConvertible {
  public var description : String {
    var desc = "TestUnion."
    switch self {
    case .data(let val): desc += "data(val: \(val))"
    case .message(let val): desc += "message(val: \(val))"
    case .success(let val): desc += "success(val: \(val))"
    }
    return desc
  }
}

extension TestUnion : Hashable {
  public var hashValue : Int {
    let prime = 31
    var result = 1
    switch self {
    case .data(let val): result = prime &* val.hashValue
    case .message(let val): result = prime &* val.hashValue
    case .success(let val): result = prime &* val.hashValue
    }

    return result
  }
}

extension TestUnion : TStruct {

  public static var fieldIds: [String: Int32] {
    return ["data": 1, "message": 2, "success": 3, ]
  }

  public static var structName: String { return "TestUnion" }

  public static func read(from proto: TProtocol) throws -> TestUnion {
    _ = try proto.readStructBegin()
    var ret: TestUnion?
    fields: while true {

      let (_, fieldType, fieldID) = try proto.readFieldBegin()

      switch (fieldID, fieldType) {
        case (_, .stop):            break fields
        case (1, .i32):                     ret = TestUnion.data(val: try Int32.read(from: proto))
        case (2, .string):                   ret = TestUnion.message(val: try String.read(from: proto))
        case (3, .bool):                    ret = TestUnion.success(val: try Bool.read(from: proto))
        case let (_, unknownType):  try proto.skip(type: unknownType)
      }
      try proto.readFieldEnd()
    }

    if let ret = ret {
      return ret
    }

    throw TProtocolError(error: .unknown, message: "Missing required value for type: TestUnion")  }
}
```

SirWellington commented on lib/swift/Sources/TApplicationError.swift:
```diff
@@ -0,0 +1,157 @@
+
+  public init() { }
+  
+  public init(thriftErrorCode code: Int, message: String?=nil) {
```
Might wanna space out the `String?=nil` to `String? = nil`

apocolipse replied:
I'll do another styling sweep here soon.

fumoboy007 commented:
```
Compile Swift Module 'Thrift' (37 sources)
/Users/…/Packages/Thrift-1.0.0/Sources/TSet.swift:22:70: warning: 'IndexableBase' is deprecated: it will be removed in Swift 4.0.  Please use 'Collection' instead
public struct TSet<Element : TSerializable & Hashable> : Collection, IndexableBase, ExpressibleByArrayLiteral, Hashable, TSerializable {
                                                                     ^
```

ChristopherRogers commented on compiler/cpp/src/generate/t_swift_3_generator.cc:
```diff
@@ -0,0 +1,2458 @@
+
+    string visibility = is_private ? "fileprivate" : "public";
+
+    out << indent() << visibility << " final class " << tstruct->get_name();
```
I think it would be a more natural choice to use Swift structs for Thrift structs, since they are the same thing conceptually. Did you have any particular reason for choosing reference types instead? There is a performance argument, as reference counting will be more performant if the struct contains lots of other reference types.

apocolipse replied:
Thrift allows for self-referring structs, Swift does not.  Making them final gives the same non-subclassable semantics while allowing to be self-referencing.  Just loses value semantics.

ChristopherRogers commented on lib/swift/Sources/TProtocol.swift:
```diff
@@ -0,0 +1,182 @@
+  case stop     = 0
+  case void     = 1
+  case bool     = 2
+  case byte     = 3
```
'byte' is deprecated in the IDL, so I would go with 'i8' instead. Relatedly, the read/write methods need to use `Int8` instead of `UInt8`.

ChristopherRogers commented on lib/swift/Sources/TMultiplexedProtocol.swift:
```diff
@@ -0,0 +1,47 @@
+      return try super.writeMessageBegin(name: serviceFunction,
+                                         type: messageType,
+                                         sequenceID: sequenceID)
+    default:
```
I would omit `default` and instead explicitly list the remaining cases for switch statements like this where the compiler can prove that all cases were exhausted.

apocolipse replied:
That would require redundantly duplicated code.  Semantically this is just "If call or oneway, pad the message call, otherwise forward it through"

apocolipse commented on compiler/cpp/src/generate/t_swift_3_generator.cc:
```diff
@@ -0,0 +1,2458 @@
+      out << "try handler." << tfunction->get_name() << "(";
+
+      t_struct* arg_struct = tfunction->get_arglist();
+      const vector<t_field*>& fields = arg
```
Due to the way thrift serializes method responses, using a Generic `Response<T, Error>` type pattern isn't doable since there can be multiple exception types that each have unique field ID's.  I visited this idea in development and decided against it.
Alternatively, I'm thinking of using Enum's as unions here (for both async and regular clients) to remedy unnecessary optionals here, so that the client can handle possible cases in a switch.

fumoboy007 commented on lib/swift/Sources/TSocketServer.swift:
```diff
@@ -0,0 +1,149 @@
+    NotificationCenter.default.removeObserver(self)
+  }
+  
+  func connectionAcctepted(_ socket: FileHandle) {
```
Typo.

fumoboy007 commented on compiler/cpp/src/thrift/generate/t_swift_3_generator.cc:
```diff
@@ -0,0 +1,2458 @@
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sstream>
+#include "t_oop_generator.h"
```
Doesn’t compile; needs to be `thrift/generate/t_oop_generator.h` and `thrift/platform.h`.

apocolipse commented:
@fumoboy007 @ChristopherRogers Have you all found any other outstanding issues that need addressing?


@nsuke @kdubb What final steps do we need to do to get this into the official Thrift codebase?

fumoboy007 commented:
One major issue I encountered is the async support. There doesn’t seem to be an asynchronous `TProcessor`/`TProtocol`/`TClient` API in the base/generated code.

Async support is important for high-efficiency server applications. (At least until [SR-2905](https://bugs.swift.org/browse/SR-2905) is implemented for Linux.) I can explain this in more detail if desired.

apocolipse commented:
@fumoboy007 There is `TAsyncClient` in TClient.swift as well as `TAsyncTransport/Factory` in TTransport.swift, and the generator will generate async clients properly (`--gen swift_3:async_clients`), for servers, the design I've seen so far is that they're asynchronous by creating new transports for each request for an async server, so its incumbent on the base thrift server implementation to handle async code, in which the simple socket server I've included should handle it fine, as well as the Perfect server example I posted in the readme should work as well. 

Can you elaborate on anything more that you'd like to see?

fumoboy007 commented:
Oh I didn’t know about the `async_clients` option… thanks! 😬

fumoboy007 commented:
I don’t see a `[MyService]ProcessorAsync` in the generated code. How do I handle requests in an asynchronous manner?

apocolipse commented:
@fumoboy007 so for clients you should be doing something like this:
```swift
let factory = THTTPSessionTransport.Factory(url: serviceURL)
let client = HermesAsyncClient(with: TBinaryProtocol.self, factory: factory)
```

For servers, its all about the server, no need to have an async processor
I'll need to update the Perfect HTTP Server example I have on my github for HTTP servers, but basically the setup is something like this (same as TSocketServer:
```swift
let server = try? TSocketServer(port: 9090,
                                 service: MyConcreteService(), // your implementation of thrift service
                                 inProtocol: TBinaryProtocol.self,
                                 outProtocol: TBinaryProtocol.self,
                                 processor: MyServiceProcessor.self) // generated service processor
```
Since it leverages generics, each request creates its own processor rather than re-using the same one over and over
(If you look at the TSocketServer impl, you'll see something like
```swift
let processor = Processor(service: self.serviceHandler)
processor.process(on: inproto, outProtocol: outproto)
```)

ChristopherRogers commented:
I have not. 👍

apocolipse commented:
I've created 2 separate HTTP Servers using external libraries and made them available to play with
https://github.com/apocolipse/TZewoHTTPServer  and
https://github.com/apocolipse/TPerfectHTTPServer

Their interface is identical to the TSocketServer included, so they can be quickly and easily swapped out to test differences between them.  Created as separate libs to avoid no external dependency requirement for Thrift.
Apple has announced a Server API team, so hopefully in the future we can have Swift STD type httpservers built-in, but for now this will do.

apocolipse commented:
Update:
I've modified the generated Async clients to use a new `TAsyncResult<T>` parameter for callbacks rather than Optional Return/Error tuples.   Short version of why is that those parameters are always mutually exclusive, and there's no really good way to have a non-optional return or throw pattern for async.
With the `TAsyncResult<T>`, non-optional return values are encapsulated within a throwing func in the enum, such that you can handle your do/catch block in a callback.  Consider the following example:

```swift
/// normal: public func hello() throws -> String
do {
  let world = try hello()
} catch let e { handle(e) }

/// async:   public func hello(completion: @escaping (TAsyncResult<String>) -> Void)
hello() {
  do {
     let world = try $0.value()
  catch let e { handle(e) }
}
```

fumoboy007 commented:
Woah, nice!

apocolipse commented on compiler/cpp/src/thrift/generate/t_swift_3_generator.cc:
```diff
@@ -0,0 +1,2553 @@
+/*
```
While I don't disagree, I can potentially see future problems arising when I try to introduce namespacing via modules.  
Additionally, Swift 2.x won't be supported by Apple any longer after 2.x next year, so that might make a good case to axe the old one and just call this the canonical `Swift` generator, thoughts?

apocolipse replied:
@nsuke I think I'd actually prefer to eliminate the old Swift generator and rename this to just t_swift_generator.cc, given that Apple will sunset Swift 2.x within the next 6 months.  Is this acceptable?  

My current fear with that is that the current swift lib/generator are not housed under a tag here, so anyone having a dependency may face issues.  
Anyone currently using it is almost certainly using Cocoapods and pointing to master, and they'd have a lockfile pointing to the commit they last used, so there's some certainty there, but an intermediate snapshot tag might help otherwise. 

Cocoapods does support pointing to explicit commits though too, a small How-To could be added to the README to help users in need to point to the old generator if they overly specifically need it.

nsuke replied:
Assigning simple "--gen swift" to this one sounds good to me, given the Swift 2.x situation and this beeing pure Swift implementation.

In general, though, removing is hard.
The most natural course of action would be to keep current one as an generator switch like "swift:cocoa".

For completely replacement, we need ordinal considerations for any breaking changes.
How hard it is to migrate existing users' code ?
It seems to me that anyone who extended library code (inheritance etc) would be in a huge trouble.

apocolipse commented on lib/swift/Package.swift:
```diff
@@ -0,0 +1,12 @@
+//let package = Package(
```
I'll add one, SPM generates this without any.

apocolipse commented on lib/swift/Sources/LinuxHelper.swift:
```diff
@@ -0,0 +1,34 @@
+//
+//  Created by Christopher Simpson on 8/22/16.
+//
+//
```
Will update

apocolipse commented on lib/swift/README.md:
```diff
@@ -0,0 +1,217 @@
+specific language governing permissions and limitations
+under the License.
+
+Brought to you by [FiscalNote, Inc](http://www.fiscalnote.com/)
```
If required by Apache sure, though my company would like a mention somewhere if possible, since they agreed to allow me build this on company time.

apocolipse commented:
@nsuke I've removed the attribution, will keep it in my personal Lib repo thats used strictly for SPM integration.  Fixed the error codes you mentioned above.

I'll go ahead and merge the old cocoa generator in and put it behind a flag.  Ideally, users leveraging existing Swift 2.x Cocoa bindings are using Cocoapods, and have a lock to the specific commit they're using.  This isn't always the case, but should serve at least a basic level of protection on their client/library dependency side.  From the Generator's perspective, it will be completely broken unless they checkout the most recent commit before I merge them (or change any build steps to use the new generator). 
In terms of their existing code if they're extending the library or inheriting from it, some of it may break given some of the interfaces have changed towards Value-typed interfaces.  Extensions should for the most part operate properly so long as they're not calling changed interfaces.   All of this is of course purely speculative as I've not gone through the process of trying to migrate.   Long story short however there should be facilities to hit old lib/generator when needed and given Swift is compiled, migrating code over shouldn't introduce any breaking changes or unexpected behaviors.

apocolipse commented:
Update 2/28/2017:
Old Swift generator and new merged into one,
Old generator under `cocoa` option, `promise_kit` option only available for old generator as well.

I don't have any Swift 2.x projects so i can't really test, but if anyone else wants to spot check code and test against old versions that'd be great.  I've left the old swift generator CC file in tact (moved to `*.cc..old`), and spaced out things such that it lines up with the new code line for line for comparison.

Next and last steps to make this production ready IMO is namespacing, I'm going to add an option to namespace by splitting into subdirectories.  Actually configuring projects will be left up to the developer.  I'll add some output describing how to do it in SPM, but injecting the targets into SPM or Xcodeproj are beyond the scope of Thrift so for now the only thing namespacing will do is place source files in their own namespace relative directory, generate imports when necessary, and appropriately namespace types.  Any suggestions or input are welcome.

jeking3 commented:
Are these changes integrated into the "make cross" cross-language validation suite?

apocolipse commented:
@jeking3 no, any docs on whats needed for that?  I can add it in.  The old Cocoa+Swift generator was lacking here as well. (As is the Obj-C Cocoa generator)

jeking3 commented:
None that I know of.  What you can do is:

1. Make sure language support is in the Dockerfile for the ubuntu and debian builds.
2. Look for CROSS_LANGS in the top level Makefile.am.  That should lead you to the rest.
3. You will need a TestServer and TestClient implementation; you can look in test/cpp or test/perl for example implementations.  You will probably need a new test/ subdirectory for swift.
4. Make cross until it works.  If a test reliably fails (and is supposed to) put it into the known test failures file by running "test.py -U merge" from the test directory after running make cross.

Note that adding to CROSS_LANGS will also run the cross feature tests as well, which are in test/features.

gotev commented:
@jeking3 @apocolipse What it's left to be done for this pull request to be merged? I can provide some help. Is [this](https://github.com/apache/thrift/pull/1084#issuecomment-289279668) everything it's needed?

apocolipse commented:
@gotev feel free to contribute anything that was mentioned there.  Its otherwise code complete (and likely better supported than the already merged Swift2 and Cocoa/Obj-C generators), I will contribute more to tests as I can but its not high priority for me right now.

fumoboy007 commented:
> SE-0162 [ https://github.com/apple/swift-evolution/blob/master/proposals/0162-package-manager-custom-target-layouts.md ] added API for controlling where the source files for each target should be found. This allows SwiftPM to support source trees that don't conform to the standard package layout conventions.

Can we take advantage of this so that people can use the library from the official repo?

apocolipse commented:
Its just a proposal so far, i don't think it'll be implemented for Swift 3 (but i havent been keeping close track, correct me if i'm wrong)  It'll make more sense to do once Swift4 drops and SPM is beefed up.

fumoboy007 commented:
Yeah, it’s implemented in Swift 4.

![img_1999](https://user-images.githubusercontent.com/2100868/26892962-c6356fa6-4b6e-11e7-8394-1964d22226ee.PNG)

apocolipse commented:
Also, to anyone using this, I'm making a pretty decent sized change right now to how `TEnums` work with Swift Enums.  Backstory is that Swift enums are overly opinionated and don't work well if a new enum value is added to a service and the client hasn't updated.

Other languages handle this pretty easily since enums are just Ints to them, they simply read and store the value even if it doesn't match a case.  Since Swift's enums are finite and opinionated, and not backed by Ints, this posed a bit of a challenge.  I initially designed the TEnums to be RawRepresentable Int32 since it just made sense at the time, this let the library handle read/write generically and the generator only needed to declare the cases and an empty init case.  The problem arose when we saw it failing when new values came in.  Granted it was designed to do so, but we needed a way internally to gracefully handle that.

As a replacement, I'm removing the RawRepresentable/Int32 requirement on TEnums, and removing the protocol read from the library as well, they will retain a rawValue and init(rawValue:) to facilitate serialization and maintain compatibility with code that may be already using those, but they're no longer inherently RawRepresentable.  The reason for this decision was to add an `.unknown(Int32)` case that retains the extra value.  
This is an opt-in case, so you'll need to use the generator flag `safe_enums`, but this way you don't lose that information, and in the likely event you'll pull something that has an unknown enum value and then need to send it back to a service, you'll retain the appropriate value and maintain compatibility between server/client version mismatch.

jeking3 commented:
What needs to happen next to move this forward?

nsuke commented:
This replaces current Swift2 generator/lib with new one which supports Swift3. The current one is still kept as option `swift:cocoa`, though.

Swift 2 has been not usable with latest XCode for months now so it makes sense to me to land this to master. Reviews have been done by contributors before, although a few more cleanup might be needed.

nsuke commented on compiler/cpp/src/thrift/generate/t_swift_generator.cc:
```diff
@@ -95,19 +110,20 @@ class t_swift_generator : public t_oop_generator {
+  //                        t_type* type,
+  //                        t_const_value* value,
+  //                        bool defval = false,
+  //                        bool is_property = false);
```
Can you remove commented out code, if useless ? Otherwise please put a comment as to why we need to keep it as a comment.

nsuke commented on compiler/cpp/src/thrift/generate/t_swift_generator.cc.old:
```diff
@@ -0,0 +1,3249 @@
+/*
```
Can you remove this one if it's no longer used ?

apocolipse commented:
@nsuke requested cleanup completed

nsuke commented on compiler/cpp/src/thrift/generate/t_swift_generator.cc:
```diff
@@ -245,22 +297,30 @@ class t_swift_generator : public t_oop_generator {
+  string out_dir = get_out_dir();
+  string module_path = out_dir;
+  string name = capitalize(program_name_);
+  if (namespaced_ and !module.empty()) {
```
I just noticed that this broke Windows build.
It seems that VC++ does not support alternative keywords in our settings.
 Can you replace`and` with `&&` ? We'll see if it fixes the build.

apocolipse commented:
@nsuke hah good catch, bad habits from other languages ;)  Fixed

tbartelmess commented:
Any movement on this? I'd love to see this merged.

jeking3 commented:
I would recommend the original commenters chime in as the requested work was done; this should be SQUASHED (nobody wants to cherry-pick 84 commits) and rebased on master and if it passes CI clean, definitely good towards merging it.  @nsuke and @ChristopherRogers both commented in the past.

apocolipse commented:
@jeking3 Squashed

jorgeazevedo commented:
Any movement on this? I'd love to see this merged.

JacopoMangiavacchi commented:
I tested this PR and I see it already support also Swift 4.0 and I'm adding Swift-Nio support for easily implement also the server side.

Any update on this merge ?  Thanks a lot

apocolipse commented:
@JacopoMangiavacchi  Server side already works using something like Perfect or Zewo, I already have separate server adapters for them https://github.com/apocolipse/TPerfectHTTPServer

I'd advise making Swift-Nio support separate as well to ensure better compatibility.

jeking3 commented:
Okay, so I could merge this, but I need to know (since I've never used cocoa, swift, or any of the apple stuff really):

* Are there breaking changes?
* Are they documented?
* Is it already running in the cross test or is that future work?
* Can this be exercised in the Travis CI build environment? (make check, make cross?)
* Can I add swift support to Ubuntu Xenial and Artful?
* What minimum and maximum versions are known supported?
* Has the LANGUAGES.md file been updated?
* Has the build/docker/README.md file been updated?
* Is this supposed to replace the cocoa implementation?  Should that be removed?

I can make some of these changes as part of finalization.  Thanks.

yurikoles commented:
Hi @apocolipse,
Please consider taking into account usage of reserved keywords in *.thrift files. For example back end developers may use field with name `description` in their structures. And this same keyword then redefined in generated extension file. This causes compile errors.

apocolipse commented:
Hey @jeking3 I'll answer as best I can given i haven't done much with Thrift in a while :) my coworker has taken up much of the Thrift work with Swift internally at my company.
- No/Yes?  The compiler is written to allow for the Old Cocoa/Swift with a compiler flag, so anyone still using swift 1 or 2 should still be able to use the Cocoa style generator but will have to use a compiler flag
- I don't believe thats documented anywhere
- No tests written for this, that is future work
- Should be exercisable in Travis though no tests currently
- I'm not sure where the Swift compiler support for Xenial and Artful currently is, but should be available.
- Given backwards compatibility with Cocoa/Swift generator this should work with all Swift versions v1 through v4
- LANGUAGES.md has not been updated
- build/docker/README.md have not been updated
- The cocoa implementation is separate from this implementation.  The Cocoa compiler/libraries exist for Obj-C, and Swift 1 & 2, whereas Swift 3+ operate independently without Cocoa (as Swift 3 and up can run on Linux without Obj-C runtime/libraries).  That being said the old Swift/Cocoa compiler is "gone/removed" and there is "only 1 Swift compiler" given the old Swift/Cocoa compiler is integrated in this as a compiler flag
- Given this is a full rewrite it will not play nice with #1002, though since that PR is related to the Swift/Cocoa compiler and that has been integrated for backwards compatibility that PR can be rewritten to work with this.  My comments on that PR still stands, the style of "Namespacing" used there by prefixing class names is a Cocoa convention that is improper in Swift, This version does proper "namespacing" for Swift by putting generated files in namespaced folders which should then be imported as separate modules, adhering to Swift's module namespacing conventions, however for the Swiftv1+2/Cocoa side its an acceptable pattern.

jeking3 commented:
Based on comments here and in THRIFT-3773 I have merged this despite it being incomplete.  I opened https://issues.apache.org/jira/browse/THRIFT-4547 to track completion of this work.
